src/log.c

```diff
@@ -337,19 +342,58 @@ static void write_log(qd_log_source_t *log_source, qd_log_entry_t *entry)
 
     aprintf(&begin, end, "\n");
 
-    if (sink->file) {
-        if (fputs(log_str, sink->file) == EOF) {
-            char msg[TEXT_MAX];
-            snprintf(msg, sizeof(msg), "Cannot write log output to '%s'", sink->name);
-            perror(msg);
-            exit(1);
-        };
-        fflush(sink->file);
+    if (sink) {
+        if (sink->file) {
+            if (fputs(log_str, sink->file) == EOF) {
+                char msg[TEXT_MAX];
+                snprintf(msg, sizeof(msg), "Cannot write log output to '%s'", sink->name);
+                perror(msg);
+                exit(1);
+            };
+            fflush(sink->file);
+        }
+        if (sink->syslog) {
+            int syslog_level = level->syslog;
+            if (syslog_level != -1)
+                syslog(syslog_level, "%s", log_str);
+        }
     }
-    if (sink->syslog) {
-        int syslog_level = level->syslog;
-        if (syslog_level != -1)
-            syslog(syslog_level, "%s", log_str);
+
+    // Emit a vanflow record for "serious" log events. Ignore vanflow logs to prevent infinite loops.
+    // Do this while holding the log_event_lock. This will prevent the vanflow thread from shutting
+    // down and freeing resources while we generate the record
+    if (level->bit >= QD_LOG_WARNING && strcmp("FLOW_LOG", entry->module) != 0) {
```

Comment:
I was wondering if instead of doing `strcmp("FLOW_LOG", entry->module)`, should we store a `qd_log_module_t`
in `qd_log_entry_t`. Will that be more efficient than `strcmp` ?

Reply:
In this case my perf testing shows the strcmp overhead to be basically none. The bit check ends up short circuiting and avoiding the strcmp entirely.

However in general storing the log module in the log entry may be a general improvement for the logging module, but that would be way beyond the scope of this patch.

Reply:
> However in general storing the log module in the log entry may be a general improvement for the logging

Yes, we could do this as a different patch. Since warnings (and CRITICAL and ERROR) are not that frequent, it is ok the way it is now. 

I don't have any more questions about this PR, it is ready to ship.